bach/bach/operations/cut.py

```diff
@@ -0,0 +1,195 @@
+        max_name = f'{self.series.name}_max'
+
+        properties_df['min_adjustment'] = self._calculate_adjustments(
+            to_adjust=properties_df.all_series[min_name], compare_with=properties_df.all_series[max_name],
+        )
+        properties_df['max_adjustment'] = self._calculate_adjustments(
+            to_adjust=properties_df.all_series[max_name], compare_with=properties_df.all_series[min_name],
+        )
+
+        # need to adjust both min and max with the prior calculated adjustment
+        # this is mainly to avoid the case both min and max are equal
+        properties_df[min_name] = (
+            properties_df.all_series[min_name] - properties_df.all_series['min_adjustment']
+        )
+        properties_df[max_name] = (
+            properties_df.all_series[max_name] + properties_df.all_series['max_adjustment']
+        )
+
+        diff_min_max = properties_df.all_series[max_name] - properties_df.all_series[min_name]
+        # value used for expanding start/end bound
+        properties_df['bin_adjustment'] = diff_min_max * self.RANGE_ADJUSTMENT
+        properties_df['step'] = diff_min_max / self.bins
+
+        return properties_df
+
+    def _calculate_buckets(self, bucket_properties_df: 'DataFrame') -> 'Series':
+        """
+        calculates each bucket for all series values.
+        """
+        return self.series.copy_override(
+            name='bucket',
+            expression=Expression.construct(
+                # we need to make sure max value is on the last bucket
+                (
+                    f'case when {{}} = {{}} then {self.bins} else \n'
+                    f'width_bucket({{}}, {{}}, {{}}, {self.bins}) end'
+                ),
+                self.series,
+                Series.as_independent_subquery(bucket_properties_df[f'{self.series.name}_max']),
+                self.series,
+                Series.as_independent_subquery(bucket_properties_df[f'{self.series.name}_min']),
+                Series.as_independent_subquery(bucket_properties_df[f'{self.series.name}_max']),
+            ),
+        )
+
+    def _calculate_adjustments(
+        self, to_adjust: 'Series', compare_with: 'Series'
+    ) -> 'Series':
+        """
+        calculates adjustment when to_adjust == compare_with.
+        If both are equal, calculations based on RANGE_ADJUSTMENT will be performed
+        """
+        case_stmt = (
+            f'case when {{}} = {{}} then\n'
+            f'case when {{}} != 0 then {self.RANGE_ADJUSTMENT} * abs({{}}) else {self.RANGE_ADJUSTMENT} end\n'
+            f'else 0 end'
+        )
+        return to_adjust.copy_override(
+            expression=Expression.construct(
+                case_stmt,
+                compare_with,
+                *[to_adjust] * 3,
+            )
+        )
+
+    def _calculate_bucket_ranges(self, bucket_properties_df: 'DataFrame') -> 'DataFrame':
+        """
+        Calculates upper and lower bound for each bucket.
```

Comment:
```suggestion
        Calculates upper and lower bound for each bucket.
        Returns a DataFrame with a row for each bucket, and four columns:
         * bucket (integer 1 to N)
         * lower_bound (float)
         * upper_bound (float)
         * range (object containing both lower_bound and upper_bound)
```
Maybe it is just me, but I think having a bit more extensive comments would make reviewing (and later editting/debugging/etc.) of this code _much_ simpler. I am having a real hard time reading this at the moment